cpp/src/arrow/compute/kernels/vector_selection_filter.h

```diff
@@ -0,0 +1,39 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+
+#include "arrow/array/data.h"
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/kernels/vector_selection_internal.h"
```

Comment:
We should not include `*internal.h` from a non `*internal.h` headers because `*internal.h` aren't installed.

Reply:
`vector_selection_filter.h` is also internal — included only by `vector_selection.cc`. Should I rename it with an `internal` suffix in the name?

Reply:
But I just realized it's not needed. :smile:  This is why I made `PopulateFilterKernels` take an output param at some point. That allows me to forward-declare `SelectionKernelData` and not include the full header.

Reply:
As it's been done with `vector_selection_take.h`

Reply:
IIRC the `internal.h` suffix is important as the cmake uses it to decide which headers get installed. So if it is actually a private/internal header it should be renamed imo. But @kou would know best!

Reply:
> IIRC the `internal.h` suffix is important as the cmake uses it to decide which headers get installed. So if it is actually a private/internal header it should be renamed imo.

Correct.

Reply:
Makes sense. I'm also renaming `vector_selection_(filter|take).(h|cc)` and pushing now.